packages/mds-agency/request-handlers.ts

```diff
@@ -402,60 +401,69 @@ export const submitVehicleTelemetry = async (
     })
     return
   }
+
   const name = providerName(provider_id)
-  const failures: string[] = []
-  const valid: Telemetry[] = []
 
-  const recorded = now()
-  const p: Promise<Device | DeviceID[]> =
-    data.length === 1 && isUUID(data[0].device_id)
-      ? db.readDevice(data[0].device_id, provider_id)
-      : db.readDeviceIds(provider_id)
   try {
-    const deviceOrDeviceIds = await p
-    const deviceIds = Array.isArray(deviceOrDeviceIds) ? deviceOrDeviceIds : [deviceOrDeviceIds]
-    for (const item of data) {
-      // make sure the device exists
-      const { gps } = item
-      const telemetry: Telemetry = {
-        device_id: item.device_id,
-        provider_id,
-        timestamp: item.timestamp,
-        charge: item.charge,
-        gps: {
-          lat: gps.lat,
-          lng: gps.lng,
-          altitude: gps.altitude,
-          heading: gps.heading,
-          speed: gps.speed,
-          accuracy: gps.hdop,
-          satellites: gps.satellites
-        },
-        recorded
-      }
+    const deviceIds: Set<UUID> = await (async () => {
+      if (data.length === 1) {
+        const [telemetry] = data
 
-      try {
-        isValidTelemetry(telemetry)
-      } catch (err) {
-        logger.info(
-          `Non-critical prototype Telemetry ValidationError for ${providerName(provider_id)}. Error: ${JSON.stringify(
-            err
-          )}`
-        )
-      }
+        if ('device_id' in telemetry) {
+          const { device_id } = telemetry
 
-      const bad_telemetry: ErrorObject | null = badTelemetry(telemetry)
-      if (bad_telemetry) {
-        const msg = `bad telemetry for device_id ${telemetry.device_id}: ${bad_telemetry.error_description}`
-        // append to failure
-        failures.push(msg)
-      } else if (!deviceIds.some(item2 => item2.device_id === telemetry.device_id)) {
-        const msg = `device_id ${telemetry.device_id}: not found`
-        failures.push(msg)
-      } else {
-        valid.push(telemetry)
+          if (isUUID(device_id)) {
+            const device = await db.readDevice(device_id, provider_id)
+
+            // Map with only one entry
+            return new Set<UUID>([device.device_id])
+          }
+        }
       }
-    }
+
+      const deviceIdsWithProviderIds = await db.readDeviceIds(provider_id)
+
+      // Turn array returned to a Set for fast lookups
+      return new Set(deviceIdsWithProviderIds.map(({ device_id }) => device_id))
+    })()
+
+    const { valid, failures } = data.reduce<{
+      valid: Telemetry[]
+      failures: { telemetry: Telemetry; reason: string }[]
+    }>(
+      ({ valid, failures }, { device_id, timestamp, charge, gps }) => {
+        const telemetry: Telemetry = {
+          device_id,
+          provider_id,
+          timestamp,
+          charge,
+          gps: {
```

Comment:
Can't you do something like this:  
```
const telemetry: Telemetry = {
          device_id,
          provider_id,
          timestamp,
          charge,
          gps
}
```

Reply:
100% could, I was thinking the same thing! I imagine there was some motivation for picking it apart like this in the past (perhaps to drop extra properties that could be included), and wanted to keep it like this for now until I switch over to AJV for telemetry validation (which can strip foreign props).

Honestly, I get sketched out doing *anything* with this data at this point in the flow, because it hasn't been properly validated yet, and this seems a little safer than just assuming `gps` doesn't have something foreign in it.